Sources/Starknet/Data/StorageProof/MerkleNode.swift

```diff
@@ -0,0 +1,74 @@
+import Foundation
+
+public enum MerkleNode: Codable, Equatable {
+    case binaryNode(BinaryNode)
+    case edgeNode(EdgeNode)
+
+    public init(from decoder: Decoder) throws {
+        let binaryNodeKeys = Set(BinaryNode.CodingKeys.allCases.map(\.stringValue))
+        let edgeNodeKeys = Set(EdgeNode.CodingKeys.allCases.map(\.stringValue))
+
+        let binaryNodeContainer = try decoder.container(keyedBy: BinaryNode.CodingKeys.self)
+
+        if Set(binaryNodeContainer.allKeys.map(\.stringValue)) == binaryNodeKeys {
+            let binaryNode = try BinaryNode(from: decoder)
+            self = .binaryNode(binaryNode)
+        } else if let edgeNodeContainer = try? decoder.container(keyedBy: EdgeNode.CodingKeys.self),
+                  Set(edgeNodeContainer.allKeys.map(\.stringValue)) == edgeNodeKeys
+        {
+            let edgeNode = try EdgeNode(from: decoder)
+            self = .edgeNode(edgeNode)
+        } else {
+            let context = DecodingError.Context(
+                codingPath: decoder.codingPath,
+                // TODO: Improve error message.
+                debugDescription: "Failed to decode MerkleNode from the given data."
+            )
+            throw DecodingError.dataCorrupted(context)
+        }
```

Comment:
It seems to me this is slightly overcomplicated:
```suggestion
        if let binaryNode = try? BinaryNode(from: decoder) {
            self = .binaryNode(binaryNode)
        } else if let edgeNode = try? EdgeNode(from: decoder) {
            self = .edgeNode(edgeNode)
        } else {
            throw DecodingError.dataCorrupted(DecodingError.Context(codingPath: decoder.codingPath, debugDescription: "Failed to decode MerkleNode."))
        }
```

Reply:
And while the spec is not merged yet, do you think maybe it's worth it to reach out to starknet-specs maintainers and ask "type" field to be added? "Type" fields are used heavily in the spec overall, so this might be a reasonable ask.